File: 0079-word-search/0079-word-search.java
class Solution {
    public boolean dfs(char[][] board, String word, int wordPos, int posx, int posy) {
        if(wordPos == word.length() - 1) return board[posx][posy] == word.charAt(wordPos);
        if(board[posx][posy] != word.charAt(wordPos)) return false;
        int m = board.length, n = board[0].length;
        int[][] directions = {{1, 0}, {0, 1}, {-1, 0}, {0, -1}};

        char temp = board[posx][posy];
        board[posx][posy] = '.';
        for(int[] neighbor : directions) {
            int dx = posx + neighbor[0];
            int dy = posy + neighbor[1];

            if(0 <= dx && dx < m && 0 <= dy && dy < n) {
                if(dfs(board, word, wordPos + 1, dx, dy)) {
                    return true;
                }
            }
        }
        board[posx][posy] = temp;
        return false;
    }

    public boolean exist(char[][] board, String word) {
        int m = board.length, n = board[0].length;
        for(int i = 0; i < m; i++) {
            for(int j = 0; j < n; j++) {
                if(dfs(board, word, 0, i, j)) {
                    return true;
                }
            }
        }
        return false;
    }
}

File: 0079-word-search/0079-word-search.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        def backTrack(x: int, y: int, pos: int):
            if board[x][y] != word[pos] or pos == len(word) - 1:
                return board[x][y] == word[pos]

            temp = board[x][y]
            board[x][y] = "."
            ans = False
            for dx, dy in [[0, 1], [1, 0], [0, -1], [-1, 0]]:
                new_x = x + dx
                new_y = y + dy
                if 0 <= new_x < n and 0 <= new_y < m:
                    ans = ans or backTrack(new_x, new_y, pos + 1)
            board[x][y] = temp
            return ans

        n = len(board)
        m = len(board[0])
        for i in range(n):
            for j in range(m):
                if backTrack(i, j, 0):
                    return True
        return False